File: repository/src/main/java/by/rekuts/travelagency/repository/impl/ReviewRepositoryImpl.java
package by.rekuts.travelagency.repository.impl;

import by.rekuts.travelagency.domain.Review;
import by.rekuts.travelagency.repository.ReviewRepository;
import by.rekuts.travelagency.repository.ReviewSpecification;
import by.rekuts.travelagency.repository.Specification;
import org.springframework.stereotype.Repository;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import javax.persistence.criteria.Predicate;
import java.util.List;

@Repository
public class ReviewRepositoryImpl implements ReviewRepository {
    @PersistenceContext
    private EntityManager entityManager;

    @Override
    public void insert(Review review) {
        entityManager.persist(review);
    }

    @Override
    public void update(Review review) {
        var dbReview = entityManager.find(Review.class, review.getId());
        dbReview.setText(review.getText());
        dbReview.setUser(review.getUser());
        dbReview.setTour(review.getTour());
        entityManager.merge(dbReview);
    }

    @Override
    public void delete(int id) {
        entityManager.remove(entityManager.find(Review.class, id));
    }

    @Override
    public List<Review> getList(Specification specification) {
        final var reviewSpecification = (ReviewSpecification) specification;
        var builder = entityManager.getCriteriaBuilder();
        var criteriaQuery = builder.createQuery(Review.class);
        var root = criteriaQuery.from(Review.class);
        List<Predicate> predicates;
        predicates = reviewSpecification.getPredicates(root, builder);

        if (!predicates.isEmpty()) {
            criteriaQuery.where(
                    predicates.toArray(new Predicate[]{})
            );
        }
        return entityManager.createQuery(criteriaQuery).getResultList();
    }
}

File: repository/src/main/scala/by/rekuts/travelagency/repository/impl/ReviewRepositoryImpl.scala
package by.rekuts.travelagency.repository.impl

import java.util

import by.rekuts.travelagency.domain.Review
import by.rekuts.travelagency.repository.{ReviewRepository, ReviewSpecification, Specification}
import javax.persistence.criteria.{CriteriaBuilder, CriteriaQuery, Predicate, Root}
import org.springframework.stereotype.Repository

@Repository
class ReviewRepositoryImpl extends ReviewRepository {

  override def update(review: Review): Unit = {
    val dbReview: Review = entityManager.find(classOf[Review], review.id)
    dbReview.text_$eq(review.text)
    dbReview.user_$eq(review.user)
    dbReview.tour_$eq(review.tour)
    entityManager.merge(dbReview)
  }

  override def delete(id: Int): Unit = entityManager.remove(entityManager.find(classOf[Review], id))

  override def getList(specification: Specification[_]): util.List[Review] = {
    val reviewSpecification: ReviewSpecification = specification.asInstanceOf[ReviewSpecification]
    val builder: CriteriaBuilder = entityManager.getCriteriaBuilder
    val criteriaQuery: CriteriaQuery[Review] = builder.createQuery(classOf[Review])
    val root: Root[Review] = criteriaQuery.from(classOf[Review])
    val predicates = reviewSpecification.getPredicates(root, builder)
    if (!predicates.isEmpty) {
      criteriaQuery.where(predicates.toArray[Predicate](new Array[Predicate](predicates.size())): _*)
    }
    entityManager.createQuery(criteriaQuery).getResultList
  }
}
